### PR TITLE
[AutoDiff][Sema] Properly type-check differentiability of parameters and results in differentiable function type declarations

### LINKED ISSUE #58085: [SR-15808] [AutoDiff] Non-differentiable closure parameters not parsed correctly
|                  |                 |
   |------------------|-----------------|
   |Previous ID       | SR-15808      |
   |Radar             | None         |
   |Original Reporter | @philipturner      |
   |Type              | Bug    |

Attachment: [Download](https://user-images.githubusercontent.com/2727770/164964442-699809ff-7760-49b5-8bb4-cafcd457b000.gz)

      
   <details>
  <summary>Additional Detail from JIRA</summary>

   |                  |                 |
   |------------------|-----------------|
   |Votes             | 0         |
   |Component/s       |     |
   |Labels            | Bug        |
   |Assignee          | None      |
   |Priority          | Medium      |

   

   md5: a3df518fba66afe2f96d3bbdb0bf649a

  </details>

**Issue Description:**

The following code produces a compiler crash during SILGen because it skipped classifying the closure as \<\<error type\>\> during AST creation:

``` swift
import _Differentiation

protocol DifferentiableCollection: Sequence {}

extension DifferentiableCollection {
  func differentiableMap(_ body: @differentiable (Void) -> Void) {
    fatalError()
  }
}
```

While this code, which correctly identifies the closure as \<\<error type\>\> during AST creation, does not crash the compiler:

``` swift
import _Differentiation

protocol DifferentiableCollection: Sequence {}

extension DifferentiableCollection {
  func differentiableMap(_ body: @differentiable (Int) -> Void) {
    fatalError()
  }
}
```

The crash log is attached.

### LINKED ISSUE #58095: [SR-15818] [AutoDiff] Compiler crash for inout @noDerivative closure parameter
|                  |                 |
   |------------------|-----------------|
   |Previous ID       | SR-15818      |
   |Radar             | None         |
   |Original Reporter | @philipturner      |
   |Type              | Bug    |

Attachment: [Download](https://user-images.githubusercontent.com/2727770/164964448-7a9e5994-d270-4702-b96e-4bf1edbb8f1b.gz)

      
   <details>
  <summary>Additional Detail from JIRA</summary>

   |                  |                 |
   |------------------|-----------------|
   |Votes             | 0         |
   |Component/s       |     |
   |Labels            | Bug        |
   |Assignee          | None      |
   |Priority          | Medium      |

   

   md5: 74613197bdf79640c6596a3142e64468

  </details>

**Issue Description:**

The following code produces a crash during compilation. It was caught while testing bug fixes brought by <https://github.com/apple/swift/pull/41174> - a pull request which fixes the crash.

``` swift
import _Differentiation

typealias MyType = @differentiable(reverse) (inout @noDerivative Float, Float) -> Void

@differentiable(reverse)
func myFunc(_ x:  inout @noDerivative Float, _ y: Float) -> Void {}

print(myFunc as MyType)
```

### PR DESCRIPTION
**Pull request by philipturner**
Resolve ambiguities between lack of a return type and lack of a _differentiable_ return type. If a function returns `Void`, it returns no values and therefore returns no _differentiable_ values. This rule allows type checking to mirror how it processes conventional functions, despite `Void` being differentiable.

```swift
// Before this PR, the following declaration might crash the compiler.
typealias FuncType = @differentiable(reverse) (inout Float, Float) -> Void

// While the following function has the same signature and compiles just fine.
@differentiable(reverse)
func foo(x: inout Float, y: Float) {
  x += y
}

```

Resolves three issues, listed below. The [SR-15818](https://bugs.swift.org/browse/SR-15818) regression test was modified to avoid triggering [SR-15823](https://bugs.swift.org/browse/SR-15823), an unrelated compiler + runtime crasher.

- Resolves (partially): #56412
- Resolves: #58085
- Resolves: #58095

### CODE REVIEW

**Comment by philipturner** [comment]
CC: @AnthonyLatsis please review

**Comment by philipturner** [comment]
#41128

**Comment by philipturner** [comment]
2/100! only 98 more AutoDiff bugs to go!

File: lib/AST/Type.cpp [hunk 1]
```diff
@@ -6021,8 +6021,7 @@ TypeBase::getAutoDiffTangentSpace(LookupConformanceFn lookupConformance) {
       newElts.push_back(elt.getWithType(eltSpace->getType()));
     }
     if (newElts.empty())
-      return cache(
-          TangentSpace::getTuple(ctx.TheEmptyTupleType->castTo<TupleType>()));
```
Comment (AnthonyLatsis) [inline lib/AST/Type.cpp:6025]: We may want to add a "non-empty tuple" assertion to `TangentSpace::getTuple` to prevent this from happening elsewhere.

**Review by AnthonyLatsis — COMMENTED** [review]
Are you planning to add some tests? I am no AutoDiff expert, so I can only speak for basic stuff.

File: lib/AST/Type.cpp (hunk 1 above)
Comment (philipturner) [reply lib/AST/Type.cpp:6025]: That sounds like a good idea. I’ll change that before this PR gets merged.

**Comment by philipturner** [comment]
> Are you planning to add some tests? I am no AutoDiff expert, so I can only speak for basic stuff.

You’re right! I should add tests! All the crashers (and the incorrect error diagnostic) I cited in SR-15808 should be tested.

File: lib/AST/Type.cpp (hunk 1 above)
Comment (philipturner) [reply lib/AST/Type.cpp:6025]: Fixed.

File: lib/AST/AutoDiff.cpp [hunk 2]
```diff
@@ -366,6 +366,13 @@ GenericSignature autodiff::getDifferentiabilityWitnessGenericSignature(
   return derivativeGenSig;
 }
 
+TangentSpace TangentSpace::getTuple(TupleType *tupleTy) {
+  if (tupleTy->getElements().size() == 0) {
+    llvm::report_fatal_error("Attempted to get tangent space of empty tuple.");
+  }
```
Comment (AnthonyLatsis) [inline lib/AST/AutoDiff.cpp:372]: Note: We mainly use `assert` to stand guard of internal invariants. `llvm::report_fatal_error` is meant for something the *user* is to blame for, like passing a flag and violating its contract.

File: lib/AST/AutoDiff.cpp [hunk 3]
```diff
@@ -366,6 +366,13 @@ GenericSignature autodiff::getDifferentiabilityWitnessGenericSignature(
   return derivativeGenSig;
 }
 
+TangentSpace TangentSpace::getTuple(TupleType *tupleTy) {
+  if (tupleTy->getElements().size() == 0) {
```
Comment (AnthonyLatsis) [inline lib/AST/AutoDiff.cpp:370]: `getNumElements` is more efficient. Or you could do `tupleTy->isVoid()`, which does the check for you.

**Comment by philipturner** [comment]
The inconsistent error diagnostic requires attention in a separate pull request. It should automatically consider non-differentiable stuff as like `@noDerivative`. But, when I bypass that check, it crashes in a later compilation stage. This restriction was never outlined in the Differentiable Programming Manifesto.

Also, there's a runtime crash here, where type checking should have prevented it from compiling. This runs successfully at runtime:
```swift
typealias MyType = @differentiable(reverse) (inout Float, @noDerivative Float, @noDerivative Int) -> Void

@differentiable(reverse)
func myFunc(_ x:  inout Float, _ q: @noDerivative Float, _ y: @noDerivative Int) -> Void {
  x
}
print(myFunc as MyType)
```
While this crashes, by adding an extra line that does nothing:
```swift
typealias MyType = @differentiable(reverse) (inout Float, @noDerivative Float, @noDerivative Int) -> Void

@differentiable(reverse)
func myFunc(_ x:  inout Float, _ q: @noDerivative Float, _ y: @noDerivative Int) -> Void {
  x
}
typealias MyType2 = @differentiable(reverse) (inout Float, @noDerivative Float, @noDerivative Int) -> Void
print(myFunc as MyType)
```
I got this crash by narrowing down the wierd error diagnostic, and I think they're related.

**Comment by AnthonyLatsis** [comment]
This one-liner crashes during type-checking on my build:
```swift
import _Differentiation

@differentiable(reverse) func myFunc(_ x:  inout Float) -> Void {}
```

**Comment by philipturner** [comment]
I don't see the crash happening. Did you build from my branch, or from a previous toolchain? The comment above was tested on the January 9, 2022 toolchain, not this branch.

**Comment by AnthonyLatsis** [comment]
> Did you build from my branch, or from a previous toolchain?

No, this is a 2-day-old trunk.

**Comment by philipturner** [comment]
Please try branch then. Or, manually insert my changes onto main (there aren't that many).

**Comment by CodaFi** [comment]
>  Void should never have been differentiable in the first place. 

Is this true? It's odd, yet a reasonable thing to take the derivative of the following enum type

```
enum Top: Differentiable {
  case one
}
```

Which is isomorphic to Void. Are we missing a builtin conformance to Differentiable?

**Comment by philipturner** [comment]
Top doesn't have a tangent vector. Try using it as the only input and output of a `@differentiable` function.

**Comment by CodaFi** [comment]
> Top doesn't have a tangent vector.

Yes it does. In fact, it's `()`, an (arbitrary) representation of a zero-dimensional vector.

> Try using it as the only input and output of a @differentiable function.

There is only one such function and it is differentiable (in the mathematical sense) because it's necessarily constant. Consider that the definition of differentiation involves quantifying over points not equal to whatever `()` is represented by in your equivalent manifold. There... aren't any of those.

**Comment by philipturner** [comment]
Then how else do you suggest we resolve this bug?

**Comment by philipturner** [comment]
Swift classifies `@differentiable (Void) -> Void` as an error.

**Comment by CodaFi** [comment]
This seems like a policy choice. Do you know what Julia or PyTorch do here?

**Comment by philipturner** [comment]
I don't think that's relevant if the Swift compiler already has treated it as non-differentiable for years. All unit tests pass with the branch how it is, without modifications. Also, you could manually conform your enumeration to Differentiable. Doing that for Void is a compile failure because tuples can't conform to protocols. Furthermore, if Void was truly differentiable, every differentiable function that returns nothing and has an inout parameter would return two semantic results (a compile failure).

**Comment by CodaFi** [comment]
> because tuples can't conform to protocols. 

We have begun to build the machinery for this with [BuiltinProtocolConformance](https://github.com/apple/swift/blob/d146b69ccc035f7ad7365bc2c69f1b0049069316/include/swift/AST/ProtocolConformance.h#L1001). It's okay for us to declare this case as unsupported, but there needs to be a diagnostic for it - I already see one for differentiable enums.

> Furthermore, if Void was truly differentiable, every differentiable function that returns nothing and has an inout parameter would return two semantic results (a compile failure).

That's a solid point, but autodiff already rejects functions with multiple result types like this. I'm not asking for that to change, rather to consider that the reasoning behind us rejecting differentiable constant functions in `R^{0}` has to be stronger than looking at the implementation and declaring it moot, or merely getting the tests to pass.

**Comment by philipturner** [comment]
Could you give me more context on how we might implement a special diagnostic for Void? For example, would we say something like:
```
Differentiation of R^{0} types coming in a future version of Swift
```

**Comment by philipturner** [comment]
> the reasoning behind us rejecting differentiable functions in R^{0} has to be stronger than looking at the implementation and declaring it moot, or merely getting the tests to pass.

We aren't changing anything new about the implementation, so I don't see why it's necessary to block this PR until it's justified. The reason there was an option for it be differentiable was probably given little thought 2 years ago (maybe an accident), along with a bunch of other commits (one of which [produces an incorrect derivative](https://github.com/apple/swift/blob/94ed352c5389f7f22b8fa1935d7e56c92e9eba7a/stdlib/public/Differentiation/FloatingPointDifferentiation.swift.gyb#L271) in the stdlib that bypasses tests to this day).

**Comment by philipturner** [comment]
Is there any proper way to add runtime crashes to Swift tests? I'm investigating the very peculiar bug from [this comment](https://github.com/apple/swift/pull/41174#issuecomment-1029293249) and I think investigation will take a long time.

**Comment by philipturner** [comment]
Hold on about merging; I just discovered a new bug that got fixed by this PR. I need to report it to JIRA and add a compiler crasher test for it. It's completely unrelated to anything we've seen with this PR.

**Comment by philipturner** [comment]
SR-15818

**Comment by philipturner** [comment]
I'm ready to effectively collaborate with the Swift Core Team on fixing bugs and stop dismissing everyone's ideas. Going silent on each other isn't a way for anyone to fix anything. I realized my mistakes and will stop misconducting myself on GitHub and Swift Forums (for real this time). The sooner this gets past us, the less time we waste.

@CodaFi please elaborate on the best way to address classifying Void as differentiable. Or whether I should close this pull request because it goes against Swift AutoDiff's theoretical foundations. Maybe the opinion of someone like @rxwei might be most appropriate here.

@BradLarson I'm looking forward to collaborating with you on fixing the remaining bugs in the differentiation transform. Although I still lack a lot of knowledge, I am ready to tackle this with an open mind. I am aiming to fix all the bugs blocking S4TF by Swift 5.7, because after that the dev branch will diverge from release again. If I fail, Google Colab and iOS will be stuck on buggy AutoDiff until February 2023. I may not achieve this deadline because it's a complicated process that requires a lot of validation, but I will make a worthy attempt. 

Note that I previously conflated the goal of fixing bugs with approving AutoDiff for Swift Evolution. My priority is bringing back Swift for TensorFlow, not contributing to the Swift Open Source Project. With this clarified, we should be able to understand what and why I contribute going forward.

**Comment by CodaFi** [comment]
> please elaborate on the best way to address classifying Void as differentiable.

I think we can support it given the machinery in the compiler today. You'll need to extend a builtin conformance to these types. It's a lot of plumbing, but taking a look at how we do this for Sendable should give you a feel for how to do this for non-nominal types in general.

**Comment by philipturner** [comment]
> I think we can support it given the machinery in the compiler today.

That machinery would take a non-negligible amount of time to make. I don't see it worth my time to complete that task. Maybe in a separate pull request initiated by a long-term member of the Swift Core Team.

How about a compromise where I remove @AnthonyLatsis's suggestions that explicitly classify `Void` as non-differentiable, for the moment? Maybe a slightly different way of solving the bug, which checks if it's `Void` outside the function call. Then, comment a TODO to patch it later. That would at least stop the compiler from crashing for the time being. 

Or at the very minimum, crash with style and emit a fatal error that explains why it just crashed. Force this crash during AST instead of SILGen to help with tracing future occurrences of it. There's a second change on this PR (`inout` differentiability) that stops a crash independent of `Void` differentiability. We could change this PR's title to focus on the second change, then merge it.

**Comment by philipturner** [comment]
There's a way to solve these crashes without making `Void` un-differentiable.
